solanaj/src/main/java/org/p2p/solanaj/kits/transaction/parser/OrcaSwapInstructionParser.kt

```diff
@@ -0,0 +1,187 @@
+                blockTime = transactionRoot.blockTime,
+                slot = transactionRoot.slot,
+                fee = transactionRoot.meta.fee,
+                source = source,
+                destination = destination,
+                amountA = sourceAmount,
+                amountB = destinationAmount,
+                mintA = null,
+                mintB = null,
+                alternateSource = sourceInfo?.destination,
+                alternateDestination = destinationInfo?.source,
+            )
+        )
+    }
+
+    private fun parseFailedInstruction(
+        signature: String,
+        transactionRoot: ConfirmedTransactionRootResponse,
+    ): Result<SwapDetails> {
+        val postTokenBalances = transactionRoot.meta.postTokenBalances
+            ?: return parseResultFailure("($signature) meta.postTokenBalances is null")
+
+        val approveInstruction = transactionRoot.transaction
+            ?.message
+            ?.instructions
+            ?.find { it.parsed?.type == "approve" }
+            ?: return parseResultFailure("($signature) transaction.message.instructions doesn't contain 'approve' type")
+
+        val sourceMint = postTokenBalances.firstOrNull()?.mint
+            ?: return parseResultFailure("($signature) postTokenBalances[0].mint aka source is null")
+        val destinationMint = postTokenBalances.lastOrNull()?.mint
+            ?: return parseResultFailure("($signature) postTokenBalances[-1].mint aka destination is null")
+
+        val source = approveInstruction.parsed?.info?.source
+        val destination = approveInstruction.parsed
+            ?.info
+            ?.owner.takeIf { destinationMint == "Ejmc1UB4EsES5oAaRN63SpoxMJidt3ZGBrqrZk49vjTZ" }
+
+        val sourceAmountAsString = approveInstruction.parsed?.info?.amount
+        val sourceAmount = sourceAmountAsString ?: postTokenBalances.firstOrNull()?.uiTokenAmountDetails?.amount ?: "0"
+        val destinationAmount = postTokenBalances.lastOrNull()?.uiTokenAmountDetails?.amount ?: "0"
+
+        return Result.success(
+            SwapDetails(
+                signature = signature,
+                blockTime = transactionRoot.blockTime,
+                slot = transactionRoot.slot,
+                fee = transactionRoot.meta.fee,
+                source = source,
+                destination = destination,
+                amountA = sourceAmount,
+                amountB = destinationAmount,
+                mintA = sourceMint,
+                mintB = destinationMint,
+                alternateSource = null,
+                alternateDestination = null
+            )
+        )
+    }
+
+    /**
+     * Check liquidity to pool
+     */
+    private fun isLiquidityToPool(innerInstructions: List<InnerInstructionDetailsResponse>?): Boolean {
+        val instructions = innerInstructions?.firstOrNull()?.instructions
+        return instructions?.size == 3 &&
+            instructions[0].parsed?.type == "transfer" &&
```

Comment:
and this names too into variables or in constants

Reply:
that's a complicated question.
You see, this parsing implementation is the same as iOS and Web one, and I think we need to keep all this code unchanged to understand parsing no matter the platform.

Reply:
hmmmm okay =(